### PR TITLE
[MINOR BC] [4.x] Fix FilesystemTenancyBootstrapper discarding configured cache and session paths

### PR DESCRIPTION
**Pull request by lukinovec**
> The cache part of this is specific to `file`-driver stores that are listed in `tenancy.cache.stores`, while `tenancy.filesystem.scope_cache` is set to `true`. The session part applies to the `file` session driver, while `tenancy.filesystem.scope_sessions` is set to `true`.
>
> Ran into this while checking whether we could drop the separate 'parallel' cache store from our boilerplate's testing setup and instead just give the 'file' store a per-process path (`framework/cache/data_<parallel testing token>`), so each test process gets its own cache directory. Turns out `scopeCache()` discards configured paths entirely, so that has no effect (see below).

Using a different directory for the `file` cache store by setting `cache.stores.file.path` (either using `config([...])`, or directly in `config/cache.php` -- doesn't matter) has no effect -- `FilesystemTenancyBootstrapper::scopeCache()` ignores `path`/`lock_path` entirely and rewrites both to a hardcoded `<storage>/framework/cache/data` path on every `tenancy()->initialize()`/`tenancy()->end()`:
```php
// In `FilesystemTenancyBootstrapper::scopeCache()` (called both in `bootstrap()` and in `revert()`)
foreach ($stores as $name) {
    $path = $storagePath . '/framework/cache/data';
    $this->app['config']["cache.stores.{$name}.path"] = $path;
    $this->app['config']["cache.stores.{$name}.lock_path"] = $path;
    ...
}
```

Specific issues with hardcoding the path like this:
- a store with a configured (non-default) path gets scoped to use the default one (what I described above)
- `lock_path` is always overwritten with `path`, so a store with a separate lock directory loses that separation
- `revert()` runs the same code, so it doesn't restore what the store was configured with before tenancy initialized -- it just re-applies the same hardcoded default. Central cache ends up using the wrong path after ending tenancy.

**`scopeSessions()` has the same bug.** It never reads `session.files`, it hardcodes `<storage>/framework/sessions` on both bootstrap and revert. So a configured session path gets discarded when tenancy initializes, and it doesn't get reverted back to what it was when tenancy ends. For example, with `session.files` set to `/tmp/foo-sessions`:

```
In tenant context:    session.files = .../storage/tenant<key>/framework/sessions
After ending tenancy: session.files = .../storage/framework/sessions
```

So after ending tenancy, sessions don't go back to the configured `/tmp/foo-sessions`. They use the hardcoded `<storage>/framework/sessions` path, which was never configured anywhere.

### The fix

In `bootstrap()`, `scopeCache()` captures the original configured paths and scopes those instead of using a hardcoded default.

- If the configured path is under the central storage path, that central part gets swapped for the tenant's storage path, keeping everything after it the same (e.g. `storage/framework/cache/data` becomes `storage/tenant1/framework/cache/data`).
- If the path isn't `storage_path()`-based, there's nothing to swap, so the tenant's suffix just gets appended to the end of the path instead.

On `revert()`, `scopeCache(false)` puts the captured paths back into `cache.stores.{$name}.path`/`lock_path` and into the resolved store instance, so central cache uses the path it was configured with again.

> The paths are captured just once, during `scopeCache()` at `bootstrap()`. If `bootstrap()` fails after `scopeCache()` (e.g. when `scopeSessions()` can't create its directory), `revert()` never runs and the config is left with the scoped paths, so capturing a second time would lose the central ones. Also, `revert()` iterates the stores whose paths were captured during bootstrap rather than `config('tenancy.cache.stores')`. Removing a store from that config in tenant context would otherwise make `revert()` skip it and leave it stuck with a tenant-scoped path, and adding one would make `tenancy()->end()` throw because its path was never captured (see the 'scopeCache ignores changes to tenancy.cache.stores made in tenant context' test). These are edge cases most users wouldn't notice, but still, worth mentioning.

`lock_path` stays `null` when a store doesn't configure it, rather than us making it default to the scoped path -- `FileStore` already falls back to `path` for locks in that case, so we can just respect the store's original config.

`scopeSessions()` does the same for `session.files`. It captures the configured path during `bootstrap()` (once, for the same reason as above), scopes it, and puts it back on `revert()` (the default `storage/framework/sessions` still ends up as `storage/tenant1/framework/sessions`, so nothing changes for the default config).

Also added tests that cover each of the issues above (+ a test for handling paths that aren't `storage_path()`-based, and one for a custom `session.files` path).

**POSSIBLE MINOR BC:** Someone with `'path' => '/var/cache/foobar'` currently gets tenant cache in `storage/tenant1/framework/cache/data`. After this fix, they get `/var/cache/foobar/tenant1`, so whatever is already cached in the old directory is orphaned. The same applies to a non-default `storage_path()`-based path, e.g. `'path' => storage_path('framework/cache/data_' . env('TEST_TOKEN', 'default'))`. The config is now respected while scoping. The same goes for sessions -- with a non-default `session.files`, tenant sessions move from `storage/tenant1/framework/sessions` to the configured path scoped for the tenant, so the sessions in the old directory are orphaned.

## Note about directory separators

While implementing a method that centralizes scoping a path to the tenant (`tenantScopedPath()`), we looked into which code should use `DIRECTORY_SEPARATOR` instead of plain `/` (for Windows compatibility, overall correctness and consistency).

In short, the separators only matter in code that compares paths -- there, both sides of the check have to use the same separators (it doesn't matter whether that's `DIRECTORY_SEPARATOR` or `/`). Strings that only get passed to the filesystem are fine with plain `/` -- the filesystem handles these just fine (for example, Laravel uses `storage_path('framework/cache/data')` as the default `file` cache store's path, and that works just fine on Windows).

A thing related to this are the `rtrim()` calls. In `diskRoot()`'s "disk present in `tenancy.filesystem.disks`, but not in `tenancy.filesystem.root_override`" code branch, `rtrim()` only trimmed the `/` separator. So if a Windows user used a local disk like that, and configured that disk's path to use a trailing separator, the method would set the disk root to a path like `C:\app\uploads\/tenant1`. In practice, this shouldn't be an issue, but trimming both `/` and `\` prevents that code from setting the root to a weird path like that (so a very low impact change).

The `tenantStoragePath()` method got the same treatment as `diskRoot()` mentioned above: the original storage path _could_ be configured with a trailing slash. Again, this was a non-issue, but only because the method's output didn't get compared to other strings in a way where this _could_  be an issue. The `rtrim` there is a _slight_ improvement, but primarily, this got changed for consistency with the change in `diskRoot()`.

## Bug Fixes

* Improved isolation of file-based caches and sessions across tenant contexts.
* Preserved central cache and session locations when switching contexts.
* Scoped cache and lock directories independently, including custom paths.
* Restored original cache and session settings after leaving a tenant context.
* Avoided scoping unsupported, missing, or newly added cache stores.
* Improved compatibility for custom filesystem paths across operating systems.

## Tests

* Expanded coverage for tenant isolation, custom directories, lock paths, session files, and restoration scenarios.

### CODE REVIEW

**Comment by lukinovec** [comment]
@coderabbitai full review

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 1]
```diff
@@ -206,15 +217,60 @@ public function scopeCache(string|false $suffix): void
...
+
             $this->app['config']["cache.stores.{$name}.path"] = $path;
-            $this->app['config']["cache.stores.{$name}.lock_path"] = $path;
+            $this->app['config']["cache.stores.{$name}.lock_path"] = $lockPath;
 
             /** @var \Illuminate\Cache\FileStore $store */
             $store = $this->app['cache']->store($name)->getStore();
             $store->setDirectory($path);
-            $store->setLockDirectory($path);
+            $store->setLockDirectory($lockPath);
```
[the remark this answers is not in the document]
Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: I wonder if this makes sense? Why would we validate the path actually? I think if your file-based cache store's path is unconfigured, you'll get errors in Laravel, loud errors. Should Tenancy even deal with misconfiguration like this? `lock_path` is skipped when missing specifically because Laravel handles null/missing `lock_path` -- it just automatically uses `path` instead. The skipping there makes sense. Unconfigured `path` is just bad configuration. So, is there a REAL reason why the bootstrapper should even consider `path` being null/unconfigured?

**Comment by lukinovec** [comment]
@coderabbitai full review

**Comment by lukinovec** [comment]
@coderabbitai full review

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 2]
```diff
@@ -191,44 +194,81 @@ public function scopeCache(string|false $suffix): void
...
-        });
+            $path = $suffix ? $this->tenantScopedPath($this->originalCachePaths[$name], $suffix) : $this->originalCachePaths[$name];
+
+            // Unlike path, lock_path is optional -- if it's not set, FileStore::lock() falls back to path
+            // itself (see `$this->lockDirectory ?? $this->directory` in FileStore). Leave it null here rather
+            // than hardcoding it to $path ourselves, so a store that didn't configure a separate lock_path
+            // doesn't end up with one.
+            $lockPath = $this->originalCacheLockPaths[$name];
+            if ($suffix && $lockPath !== null) {
+                $lockPath = $this->tenantScopedPath($lockPath, $suffix);
```
[the remark this answers is not in the document]
Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php:225]: I think this suggestion is exactly the same one as you already submitted in your past reviews, and it always got dismissed.

We'll disregard this since updating scope_cache or scope_sessions in tenant context is just extremely unlikely, unrealistic thing to do.

File: tests/SessionSeparationTest.php [hunk 3]
```diff
@@ -83,6 +83,57 @@
...
+        'tenancy.bootstrappers' => [FilesystemTenancyBootstrapper::class],
+        'session.driver' => 'file',
+        'session.files' => $configuredSessionPath,
+    ]);
+
+    $sessionPath = fn () => invade(app('session')->driver()->getHandler())->path;
+
+    expect($sessionPath())->toBe($configuredSessionPath);
+
+    File::cleanDirectory($configuredSessionPath); // clean up the configured sessions dir from past test runs
```
[the remark this answers is not in the document]
Comment (lukinovec) [reply tests/SessionSeparationTest.php:100]: See the test above ('file sessions are separated'). There's just `File::cleanDirectory(storage_path("framework/sessions"))`, no `File::ensureDirectoryExists(...)` before that. Any reason why omitting `File::ensureDirectoryExists(...)` would be an issue in the currently reviewed test, and in the test above it, it's never been an issue?

**Comment by lukinovec** [comment]
### Possible further improvement (not worth pursuing, keeping this here for future reference)

Rather than `tenantScopedPath()` hardcoding how a store's path gets scoped, that could go through config, the same way `diskRoot()` resolves `root_override` templates. Something like:

```php
// config/tenancy.php
'cache' => [
    'path_override' => [
        'file' => '%configured_path%/%suffix%',
    ],
],
```

with `%configured_path%`, `%suffix%`, `%storage_path%` and `%original_storage_path%` placeholders. Stores without an entry would keep the behavior described above (no template involved).

**What that would let people do.** Here's the default `file` store from `config/cache.php`:

```php
'file' => [
    'driver' => 'file',
    'path' => storage_path('framework/cache/data'),
    'lock_path' => storage_path('framework/cache/data'),
],
```

The changes in this PR scope that path to `storage/tenant1/framework/cache/data`, so each tenant's cache directory sits next to their files:

```
storage/
├── framework/cache/data/      // central cache
├── tenant1/
│   ├── app/
│   └── framework/cache/data/
└── tenant2/
    ├── app/
    └── framework/cache/data/
```

With `'file' => '%configured_path%/%suffix%'`, the cache directory would stay where it's configured and each tenant would get a subdirectory in it:

```
storage/
├── framework/cache/data/
│   ├── tenant1/
│   └── tenant2/
├── tenant1/
│   └── app/
└── tenant2/
    └── app/
```

All tenant cache directories in one place, so anything that only concerns cache (skipping it in backups, deleting old files) is one path instead of one per tenant. And apparently, it makes mounting possible -- you can't mount tmpfs on `storage/tenant3/framework/cache/data` before tenant3 exists, but you can mount it once on `storage/framework/cache/data` and every tenant's cache ends up inside the mount.

But you can already get the same thing without a new config key. If the store's path is outside `storage_path()`, the changes in this PR append the suffix to it automatically:

```php
'path' => '/var/cache/myapp',   // in tenant1's context, this would become '/var/cache/myapp/tenant1'
```

So `path_override` would only really add one thing -- the same structure while the cache stays inside `storage/`. That does matter to someone whose deploy scripts, `.gitignore` and volume mounts all assume `storage/`, but that's about it.

**So I don't think we should add it.** AFAIK, nobody has ever asked for a configurable cache path (I searched the issues and PRs and couldn't find anything that would suggest that this is something people want). There are also two problems with the idea itself:

- `path` and `lock_path` would go through the same template, so `'%storage_path%/framework/cache/data'` brings back the first two bugs from the list above -- the configured path gets thrown away, and both directories end up in the same place. We'd have to require the template to contain `%configured_path%`.
- The placeholders wouldn't match `root_override`, where the tenant part is `%tenant%` (the tenant key). Here it'd be `%suffix%` (`suffix_base` + key), so two names for nearly the same thing in one class. I think that could be confusing.

**Comment by lukinovec** [comment]
For the record, https://github.com/archtechx/tenancy/pull/1473#pullrequestreview-4983685009 is the same suggestion about releasing the locks as coderabbitai suggested like 4 times before. Since we delete the directories in the tests, I don't consider these suggestions valid, and won't address them.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 4]
```diff
@@ -94,7 +97,7 @@ protected function storagePath(string|false $suffix): void
 
     protected function tenantStoragePath(string $suffix): string
     {
-        return $this->originalStoragePath . "/{$suffix}";
+        return $this->originalStoragePath . DIRECTORY_SEPARATOR . $suffix;
```
[the remark this answers is not in the document]
Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: > Add a Windows regression test.

No, the test suite runs on Linux

> tenantScopedPath() normalizes configuredPath but not originalStoragePath

Note that the normalization is there only to handle a rare edge case where someone configures a path built using concatenation, e.g. `storage_path() . /framework/foo_sessions`. We do use that in the tests, but it's a pretty rare edge case.

> On Windows, the prefix comparison can fail, causing cache and session paths to use the external-path branch and place tenant paths under the central storage path

For this to be a problem, someone would have to configure their storage path to use `/` *directly*, which sounds even more unlikely than the normalization use case mentioned above.

Nothing stops a user from doing that, we already handle a *similar* edge case, and this would probably require us to add just 1 similar LOC in `tenantScopedPath` that would fix an asymmetry. So I lean towards implementing this.

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 5]
```diff
@@ -191,44 +194,83 @@ public function scopeCache(string|false $suffix): void
...
+        $storagePath = str_replace('/', DIRECTORY_SEPARATOR, $this->originalStoragePath);
+
+        if (str_starts_with($configuredPath, $storagePath . DIRECTORY_SEPARATOR)) {
+            // Swap the central storage path prefix for the tenant's.
+            // For example, storage_path('framework/cache/data') becomes storage_path('tenant1/framework/cache/data').
+            return str($configuredPath)
+                ->after($storagePath . DIRECTORY_SEPARATOR)
+                ->prepend($this->tenantStoragePath($suffix) . DIRECTORY_SEPARATOR)
+                ->toString();
+        }
```
Comment (stancl) [inline src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Since we discussed that normalizing directory separators is only needed for checks, is there a reason why we do things like this instead of just in the check itself?

I think on Windows this would end up turning all storage based paths such as `storage_path('foo/bar')` used in any config files to `C://foo/bar` unnecessarily when we could keep the original separators and that way intervene less.

It doesn't really matter for functionality but it seems to me we're doing *more* here than needed which even if it doesn't cause issues in the future could just make this code harder to understand.

Specifically I'm wondering if we could just do
```php
if (str($configuredPath)->rtrim('/\\')->startsWith(rtrim($this->originalStoragePath, '/\\'))
```
or something similar.

If we also want to handle cases where `storage_path()` may be defined as `foo/bar` but then for whatever reason `$configuredPath` is `foo\bar/baz` then we can simply use `replace()` *in the check* instead of rtrim. But still don't see why we'd use the replaced/normalized strings for the actual value we return.

Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Good call. Fixed and updated the method in https://github.com/archtechx/tenancy/pull/1473/commits/127945190ee60cba00e81dbf6bab3133495428a3

> If we also want to handle cases where storage_path() may be defined as foo/bar but then for whatever reason $configuredPath is foo\bar/baz then we can simply use replace() in the check ...

Though note that this case is not covered -- paths like this won't pass the `str($configuredPath)->startsWith...` check.

I'll see how exactly would the code that would treat this edge case look like, and let you know if I think it's worth dealing with at all

Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: The code would then have to look like this:
```php
$centralPath = rtrim($this->originalStoragePath, '/\\');
$normalize = fn (string $path) => str_replace('/', DIRECTORY_SEPARATOR, $path);

if (str($normalize($configuredPath))->startsWith($normalize($centralPath) . DIRECTORY_SEPARATOR)) {
    return str($configuredPath)
        ->substr(mb_strlen($centralPath))
        ->prepend($this->tenantStoragePath($suffix))
        ->toString();
}
```

So that's not a simple `replace()`, it brings back the normalization that I got rid in the commit linked above, and in the check's body, we have to work with the not-normalized path. I find that more complex than necessary, since the edge case you described seems very unlikely.

Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Since code like that would be problematic, we agreed that if there's no better solution here, we should just revert this function to its pre-[127945190ee60cba00e81dbf6bab3133495428a3](https://github.com/archtechx/tenancy/pull/1473/changes/127945190ee60cba00e81dbf6bab3133495428a3) state. The only trade-off there is that we'll be returning the normalized path.

Reverted in https://github.com/archtechx/tenancy/pull/1473/commits/0461697e6d7804863e874c49b84f20d3c98df0d7

Comment (lukinovec) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Follow-up: https://github.com/archtechx/tenancy/pull/1473/commits/189c2b7436371f087bc32f432bc042f476b1bc44

File: src/Bootstrappers/FilesystemTenancyBootstrapper.php [hunk 6]
```diff
@@ -191,44 +194,102 @@ public function scopeCache(string|false $suffix): void
...
+        $path = str_replace('/', DIRECTORY_SEPARATOR, $path);
+
+        // On Windows, a path starting with two separators is a UNC path (e.g. '\\server\share'),
+        // so the leading separator that got collapsed by deduplicate() should be added back.
+        $uncPrefix = DIRECTORY_SEPARATOR === '\\' && str_starts_with($path, '\\\\') ? DIRECTORY_SEPARATOR : '';
+
+        // Since deduplicate() only supports UTF-8 paths, paths with non-UTF-8 characters will keep the duplicate separators.
+        $path = str($path)->deduplicate(DIRECTORY_SEPARATOR)->toString() ?: $path;
+
+        return $uncPrefix . rtrim($path, DIRECTORY_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (stancl) [reply src/Bootstrappers/FilesystemTenancyBootstrapper.php]: Does https://github.com/archtechx/tenancy/pull/1473/commits/9cf2444803795e4d312ce555688305a3ffe7dc9d fix this?

**Review by stancl — APPROVED** [review]
